File: com.stansassets.android/Samples/CallbackSample.cs
using System;
using StansAssets.Android;
using UnityEngine;

[Serializable]
public class CallbackDataSample
{
    [SerializeField]
    string m_StringData = default;

    [SerializeField]
    int m_IntData = default;

    public int IntData => m_IntData;
    public string StringData => m_StringData;
}

public static class CallbackSample
{
    const string k_FullJavaClassName = "com.stansassets.android.samples.CallbackSample";

    public static void ModelCallbackSample(int x, int y, string name, Action<CallbackDataSample> callback)
    {
        var requestBuilder = new JavaRequestBuilder(k_FullJavaClassName, nameof(ModelCallbackSample));
        requestBuilder.AddArgument(x);
        requestBuilder.AddArgument(y);
        requestBuilder.AddArgument(name);
        requestBuilder.AddAction(callback);
        requestBuilder.Invoke();
    }
}

File: AndroidStudioProject/unity/src/main/java/com/stansassets/android/samples/CallbackSample.java
package com.stansassets.android.samples;

import com.stansassets.android.IUnityCallback;
import com.stansassets.android.UnityBridge;

public class CallbackSample {

    public static class CallbackDataSample {
        String m_StringData;
        int m_IntData;
    }

    public static void modelCallbackSample(int x, int y, String name, final IUnityCallback callback) {
        CallbackDataSample dataSample = new CallbackDataSample();

        dataSample.m_StringData = "My String data";
        dataSample.m_IntData = 100;

        UnityBridge.registerMessageHandler();
        UnityBridge.sendCallback(callback, dataSample);
    }
}
